iree/modules/hal/hal_module.h

```diff
@@ -54,6 +54,22 @@ IREE_API_EXPORT iree_status_t
 iree_hal_module_create(iree_hal_device_t* device, iree_allocator_t allocator,
                        iree_vm_module_t** out_module);
 
+// Returns the device currently in use by the HAL module.
+// Returns NULL if no device has been initialized yet.
+IREE_API_EXPORT iree_hal_device_t* iree_hal_module_state_device(
+    iree_vm_module_state_t* module_state);
```

Comment:
This is forward-looking to #5724 where you won't have to provide a device and instead the user module will select the device itself. That's the other work I have in-flight. You'll still be able to manually specify devices (which is useful to avoid enumeration when you are targeting a single device), but both paths will be valid.